Alternative implementation of the algorithm

yakovs83 opened the pull request:
I left your current implementation in place and added new one on the side. The switch in the MazeApp is from gameStart -> gameStart2. Still needs some cleanup... There is some unused code, etc.

yakovs83 committed:
- Tried a different edge set generation approach, storing set labels and sets of cells in two separate data structures. Does not seem to be significantly faster.
- Changed rendering from rectangle filling to line drawing
- Tried immutable data structures and recursive implementation of edge removal. Horribly slow, not really an option.

yakovs83 committed:
- Tried a different edge set generation approach, storing set labels and sets of cells in two separate data structures. Does not seem to be significantly faster.
- Changed rendering from rectangle filling to line drawing
- Tried immutable data structures and recursive implementation of edge removal. Horribly slow, not really an option.

yakovs83 committed:
- Updated maze generation to be fully in the reference coordinate system
- Coordinate scaling is now performed at the rendering stage
- Added explicit maze boundary to be able to visually test cases when we have only 1 row or only 1 column
- Updated Coord type to be (Int,Int) since I assume we are dealing with pixels anyway

nhomble commented on src/main/scala/org/hombro/maze/Maze.scala:
```diff
@@ -0,0 +1,132 @@
+  type Label = Int
+  // we represent side of a cell as an ordered tuple of two integers (c1,c2)
+  // where c1 and c2 are cell ids and c1 < c2
+  type Side = (Int, Int)
```
if the Ints are meant to be CellID, then can't you pass the type directly into there?

nhomble commented on src/main/scala/org/hombro/maze/Maze.scala:
```diff
@@ -0,0 +1,132 @@
+    val cellLabels: mutable.Map[CellID, Label] = mutable.Map((0 until params.nx * params.ny).map(id => (id, id)): _*)
+    val cellSets: mutable.Map[Label, Set[CellID]] = mutable.Map()
+    val shuffled = util.Random.shuffle(edges)
+    var res: List[Side] = Nil
```
instead of reassigning, wouldn't it better to make this a buffer to make explicit the mutability?

nhomble commented on src/main/scala/org/hombro/maze/MazeApp.scala:
```diff
@@ -22,7 +22,7 @@ object MazeApp {
     dom.document.body.appendChild(canvas)
 
-    gameStart(canvas, context)
+    gameStart2(canvas, context)
```
I'd rather we add some button switch on the page so we can compare maybe the speed of the impls since I'm gonna fix mine shortly. If you don't wanna add it, then I'll do it.

yakovs83 committed:
Merge branch 'master' of https://github.com/nhomble/kruskal-maze

yakovs83 committed:
- Wrote a third implementation that uses the single starting cell and then tracks edge removal by advancing a wavefront
- Some of the commonly used functions were factored out into the package object
